cv32e40s/tests/programs/custom/pma/pma.c

```diff
@@ -13,319 +13,1147 @@
-  __asm__ volatile("lw %0, -4(%1)" : "=r"(tmp) : "r"(MEM_ADDR_1));
-  assert_or_die(tmp, 0xAAAAAAAA, "error: misaligned IO/MEM first store entered bus\n");
-  __asm__ volatile("lw %0, 0(%1)" : "=r"(tmp) : "r"(MEM_ADDR_1));
-  assert_or_die(tmp, 0xBBBBBBBB, "error: misaligned IO/MEM second store entered bus\n");
-  // TODO:INFO:silabs-robin  how to programmatically confirm that these region settings match as intended?
-
-
-/* A-ext not for 40s
-  printf("pma: 6. Atomics should work only where it is allowed\n");
-
-  printf("pma: Sanity check that atomic ops (lr/sc) to allowed regions is ok\n");
-  reset_volatiles();
-  __asm__ volatile("lr.w %0, 0(%1)" : "=r"(tmp) : "r"(MEM_ADDR_1));
-  __asm__ volatile("sc.w %0, %0, 0(%1)" : "=r"(tmp) : "r"(MEM_ADDR_1));
-  assert_or_die(g_mcause, -1, "error: atomics to legal region should not except\n");
-  assert_or_die(g_mepc, -1, "error: atomics to legal region unexpected mepc\n");
-  assert_or_die(g_mtval, -1, "error: atomics to legal region unexpected mtval\n");
-
-  printf("pma: Load-reserved to disallowed regions raises precise exception\n");
-  reset_volatiles();
-  __asm__ volatile("lr.w %0, 0(%1)" : "=r"(tmp) : "r"(IO_ADDR));
-  assert_or_die(g_mcause, EXCEPTION_LOAD_ACCESS_FAULT, "error: load-reserved to non-atomic should except\n");
-  assert_or_die(g_mepc, IO_ADDR, "error: load-reserved to non-atomic unexpected mepc\n");
-  assert_or_die(g_mtval, MTVAL_READ, "error: load-reserved to non-atomic unexpected mtval\n");
-
-  printf("pma: Store-conditional to disallowed regions raises precise exception\n");
-  reset_volatiles();
-  __asm__ volatile("sc.w %0, %0, 0(%1)" : "=r"(tmp) : "r"(IO_ADDR));
-  assert_or_die(g_mcause, EXCEPTION_STOREAMO_ACCESS_FAULT, "error: store-conditional to non-atomic should except\n");
-  assert_or_die(g_mepc, IO_ADDR, "error: store-conditional to non-atomic unexpected mepc\n");
-  assert_or_die(g_mtval, MTVAL_READ, "error: store-conditional to non-atomic unexpected mtval\n");
-*/
-
-
-  // Zc features are in the vplan, but are much easier to check with assertions.
-
-
-  printf("\nGoodbye, PMA test!\n\n");
-  return EXIT_SUCCESS;
+
+  *g_exp_fault = 0;
+
+  __asm__ volatile(R"(
+    # Check if second store op reached bus
+    lw t0, 0(%[addr])
+    add %[tmp], t0, zero
+  )"
+    : [tmp] "+r"(tmp)
+    : [addr] "r"(mem_addr_1)
+    : "t0", "memory"
+  );
+
+  test_fail += *g_exp_fault;
+  test_fail += tmp != 0xbbbbbbbb;
+  test_fail += g_mcause->raw != 0;
+  test_fail += *g_mepc       != 0;
+  test_fail += *g_mtval      != 0;
+  clear_status_csrs();
+  cvprintf(V_DEBUG, "Tmp: %08lx\n", tmp);
+
+  if (test_fail) {
+    cvprintf(V_LOW, "\nTest: \"%s\" FAIL!\n", name);
+    return index + 1;
+  }
+  cvprintf(V_LOW, "\nTest: \"%s\" OK!\n", name);
+  return 0;
+}
+
+// -----------------------------------------------------------------------------
+
+// TODO: implement the following for 40X:
+// Plan: Detect 40s/40x by checking csrs (avoid ifdefs)
```

Comment:
What the comment above mentions is possible, albeit tricky (have to use .word(machinecode) instead of asm instructions, otherwise the compiler will not compile the code for 40s

Reply:
Maybe I am misunderstanding the intend here, but this should be easy since mvendor+march CSRs should be a globally unique way of determining which core the code is running on.

Reply:
That is corect Mike, the issue is compiler flags here - because we write explicit asm-instructions, the ones that are not enabled in the compiler flags for a given core will cause a compilation failure. Correct me if I am wrong, but I think we either need separate tests, separate includes or ifdefs to solve this for both cores.

Reply:
Ahhh, yes. You are right.

Reply:
One option, albeit unreadable, is to use .word(machinecode) instead, as that will work with any assembler, but this will be a nuisance to debug (and not very feasible to do with jump/branch instructions)